src/jupyterlabcontroller/storage/k8s.py

```diff
@@ -0,0 +1,518 @@
+        )
+        await self.api.create_namespaced_secret(namespace, secret)
+        return
+
+    async def read_secret(
+        self,
+        name: str,
+        namespace: str,
+    ) -> Secret:
+        secret: V1Secret = await self.api.read_namespaced_secret(
+            name, namespace
+        )
+        secret_type = secret.type
+        return Secret(data=secret.data, secret_type=secret_type)
+
+    async def create_configmap(
+        self,
+        name: str,
+        namespace: str,
+        data: Dict[str, str],
+        immutable: bool = True,
+    ) -> None:
+        configmap = V1ConfigMap(
+            data=data,
+            immutable=immutable,
+            metadata=self.get_std_metadata(name="configmap"),
+        )
+        await self.api.create_namespaced_configmap(namespace, configmap)
+
+    async def create_network_policy(
+        self,
+        name: str,
+        namespace: str,
+    ) -> None:
+        api = client.NetworkingV1Api(self.k8s_api)
+        policy = V1NetworkPolicy(
+            metadata=self.get_std_metadata(name),
+            spec=V1NetworkPolicySpec(
+                policy_types=["Ingress"],
+                pod_selector=V1LabelSelector(
+                    match_labels={"app": "jupyterhub", "component": "hub"}
+                ),
+                ingress=[
+                    V1NetworkPolicyIngressRule(
+                        ports=[
+                            V1NetworkPolicyPort(
+                                port={"port": "http"},
+                            ),
+                            V1NetworkPolicyPort(
+                                port={"port": 8081},
+                            ),
+                        ],
+                    ),
+                ],
+            ),
+        )
+        await api.create_namespaced_network_policy(namespace, policy)
+
+    async def create_quota(
+        self,
+        name: str,
+        namespace: str,
+        quota: UserResourceQuantum,
+    ) -> None:
+        quota_obj = V1ResourceQuota(
+            metadata=self.get_std_metadata(name),
+            spec=V1ResourceQuotaSpec(
+                hard={
+                    "limits": {
+                        "cpu": str(quota.cpu),
+                        "memory": str(quota.memory),
+                    },
+                }
+            ),
+        )
+        await self.api.create_namespaced_resource_quota(namespace, quota_obj)
+
+    async def create_pod(
+        self, name: str, namespace: str, pod: PodSpec
+    ) -> None:
+        # Here's where we handle pull secrets.  We look for a secret named
+        # "pull-secret" in the target namespace, and if it exists, we jam it
+        # into the PodSpec.  That's why we had to wait for pull-secret
+        # creation in the namespace-resource-creation step.
+        pull_secret = True
+        try:
+            _ = self.api.read_namespaced_secret(
+                "pull-secret", namespace=namespace
+            )
```

Comment:
This feels a bit too clever to me. Shouldn't the caller know whether we're creating a pull secret, and if so, just tell this method that there is a pull secret?

Reply:
...maybe?  But on the other hand, we know we want a pull secret if and only if the controller has one, because what the controller is pulling is exactly the images that will become user pods, so we know its pull secret and the pull secret for user pods are the same.